### ISSUE TITLE
Document the process of exporting lua bindings

### ISSUE THREAD

**Issue by peterbell10** [cuberite/cuberite#4273]
The process by which Lua bindings are exported should be documented to help new contributors.
Things this documentation should cover:

- [ ] tolua++ header comment annotations `// tolua_begin`, `// tolua_end` and `// tolua_export`
- [ ] special defines: `TOLUA_TEMPLATE_BIND` and `TOLUA_EXPOSITION`
- [ ] API stability and how to make changes to the API
- [ ] writing manual bindings: `CheckParam`s, `GetStackValues`, `Push` and the `int` return value
- [ ] adding a manual binding to the module exports

**Comment by madmaxoft** [comment]
I'll do this.

**Comment by madmaxoft** [comment]
I've got this, what do you think?
http://xoft.cz/cuberite/docs/ExportingAPI.html

It doesn't cover `TOLUA_TEMPLATE_BIND` and `TOLUA_EXPOSITION`, since I have never actually come across those two.

If this is sufficient, I'll commit the HTML file as docs/ExportingAPI.html , next to the other documentation.

**Comment by peterbell10** [comment]
I like it a lot. It definitely covers all of the essentials, though I'm a bit confused by the signature of `fnImplementation`.  Where does `a_NumParams` come from?

Maybe also some cosmetic changes, I think it would be nice to break up the explanations with small code examples to help the reader follow along with what's described in the text.  Also, I find it awkward to read the text when it's taking up a full widescreen monitor.  Would it be possible to do it in markdown instead?

As for the macros, they are very uncommon so I would understand them being left out.  However, for the sake of completeness: `TOLUA_TEMPLATE_BIND` is how you export class templates to lua and `TOLUA_EXPOSITION` is the conventional way to write declarations that exist only to be read by the tolua++ parser. You can see both of these being used for `Vector3<T>`:
https://github.com/cuberite/cuberite/blob/950aeffff8d641a0855fbd9a9b2993d9dfd7d2c3/src/Vector3.h#L13
https://github.com/cuberite/cuberite/blob/950aeffff8d641a0855fbd9a9b2993d9dfd7d2c3/src/Vector3.h#L24-L34

NB: tolua++ doesn't actually care about `TOLUA_EXPOSITION` specifically, `#if 0` would work exactly the same.

**Comment by tigerw** [comment]
Is there any chance of ever moving to something like Swig?

**Comment by tigerw** [comment]
http://swig.org/

**Comment by madmaxoft** [comment]
> Where does `a_NumParams` come from?

You're right, I was writing this from memory and just thought the parameter is there, but it isn't.

> Would it be possible to do it in markdown instead?

That wouldn't help the widescreen issue. And no. First, there are other HTML files already in the folder, so it's consistent. Second, I don't think markdown supports even those small formatting things already used - I'm not aware of a way to link to an anchor inside the current document. Also, HTML is viewable anywhere, whereas markdown needs a specialized website such as GitHub to be rendered properly.

> Is there any chance of ever moving to something like Swig?

I consider that unlikely. It's much more likely that I (or someone else) will pick up the LuaStateParamsRead branch and finish it up, then get rid of the ToLua++ parser altogether and have all APIs exported manually, using the support code in that branch.
It would be actually awesome, since it could support some cool stuff as type-less and anonymous vectors (`player:teleportTo({0, 70, 0})`) and we could finally make all static API functions callable by either dot or colon, etc.

**Comment by peterbell10** [comment]
>  HTML is viewable anywhere, whereas markdown needs a specialized website such as GitHub to be rendered properly.

You could maybe try setting up [doxygen](https://www.stack.nl/~dimitri/doxygen/manual/markdown.html) to include the markdown files. There seems to be a table of contents extension as well, maybe that allows linking to sections? If not then perhaps use some css to restrict the width of the page.

> It's much more likely that I (or someone else) will pick up the LuaStateParamsRead branch and finish it up

I was having a little look into SWIG and I found this alternative called [`sol2`](https://github.com/ThePhD/sol2) that looks really interesting.  It's a very comprehensive wrapper around lua's stack api, similar to `cLuaState` but what caught my eye was this [example of overloading](https://github.com/ThePhD/sol2/blob/6638b300ccc26696afbee03a8fa08f2ddfe80d85/examples/overloading.cpp).
```cpp
sol::state lua;
lua.set_function("func", sol::overload(
        [](int x) { return x; }, 
        make_string, 
        my_add
    )
);
```

It's basically exactly what I had in mind when I commented on https://github.com/cuberite/cuberite/pull/3981#issuecomment-326142407.  What's more, according to [these benchmarks](https://sol2.readthedocs.io/en/latest/benchmarks.html), sol2 bindings are much faster than tolua++ bindings.  In fact, it seems to be almost as fast as writing the bindings by hand in plain C.

The only downside I can see is that it's written in C++14 so we would have to require newer compilers.  Thoughts?

**Comment by madmaxoft** [comment]
There's 1 issue with sol2 that could bite us, it doesn't (and cannot) support default parameter values:
https://github.com/ThePhD/sol2/issues/340

**Comment by peterbell10** [comment]
I'm not sure that's a big deal, you would just need to bind it as a a set of overloads.

I think it should also be possible to adapt tolua's code generator to output calls to sol2.  And since sol2 supports customising push, get and check then that would mean that many of today's manually bound functions could be generated automatically.
